src/components/pages/dashboard-pages/Notifications.jsx

```diff
@@ -1,10 +1,194 @@
-const Notifications = () => {
+import { useState, useEffect } from 'react';
+import { getClassesData, getAllUsersInfo } from '../../../util/DataBaseRequests';
+import { useAuth } from '../../../AuthProvider';
+
+const Notifications = ({ socket }) => {
+  const { userData } = useAuth();
+  const [classes, setClasses] = useState([]);
+  const [applicants, setApplicants] = useState([]);
+  const [classesError, setClassesError] = useState(null);
+  const [applicantsError, setApplicantsError] = useState(null);
+
+  useEffect(() => {
+    if (userData.role !== 'teacher') return;
+
+    const getTeacherClasses = async () => {
+      try {
+        const response = await getClassesData();
+        const allClasses = response.classes || [];
+        const myClassIds = userData.myClasses.map(id => id);
+
+        const filteredClasses = allClasses.filter(classInfo =>
+          myClassIds.includes(classInfo._id) &&
+          Array.isArray(classInfo.applications) &&
+          classInfo.applications.length > 0
+        );
+
+        // Sort applications within each class
+        const classesWithSortedApplications = sortClassesByApplicationDate(filteredClasses);
+
+        // Now sort the classes based on the earliest application date
+        const sortedClasses = sortClassesByEarliestApplicationDate(classesWithSortedApplications);
+
+        setClasses(sortedClasses);
+        setClassesError(null);
+      } catch (error) {
+        console.error('Error fetching classes data:', error);
+        setClassesError({ message: 'Failed to fetch classes. Please try again later.' });
+      }
+    };
+
+    getTeacherClasses();
+  }, [userData]);
+
+  useEffect(() => {
+    if (userData.role !== 'teacher' || classes.length === 0) return;
+
+    const getStudentInfo = async () => {
+      try {
+        const token = userData.token;
+        const response = await getAllUsersInfo(token);
+        const allApplicants = response || [];
+        const myApplicantIds = classes.flatMap(classInfo =>
+          classInfo.applications.map(application => application.userId)
+        );
+
+        const filteredApplicants = allApplicants.filter(response =>
+          myApplicantIds.includes(response._id)
+        );
+
+        setApplicants(filteredApplicants);
+        setApplicantsError(null);
+      } catch (error) {
+        console.error('Error fetching student data:', error);
+        setApplicantsError({ message: 'Failed to fetch students. Please try again later.' });
+      }
+    };
+
+    getStudentInfo();
+  }, [classes, userData]);
+
+  const sortClassesByApplicationDate = (classes) => {
```

Comment:
You have fantastic functionality here! Now the sorting and formatting logic can be abstracted to their own utility module. That'll clean things up a lot in this component!

Reply:
Done✅
Add NotificationsUtils file for sorting and formatting functions